vortex-serde/src/layout/mod.rs

```diff
@@ -0,0 +1,495 @@
+                Ok(Box::new(FlatLayout::new(
+                    flat_layout.begin(),
+                    flat_layout.end(),
+                    self.ctx.clone(),
+                    message_cache,
+                )))
+            }
+            LayoutVariant::NestedLayout => {
+                let nested_layout = fb_layout
+                    .layout_as_nested_layout()
+                    .expect("must be nested layout");
+                Ok(self
+                    .layout_ctx
+                    .lookup_layout_def(&LayoutId(nested_layout.encoding()))
+                    .ok_or_else(|| {
+                        vortex_err!("Unknown layout definition {}", nested_layout.encoding())
+                    })?
+                    .layout(fb_bytes, fb_loc, scan, self.clone(), message_cache))
+            }
+            _ => unreachable!("Unknown flatbuffer layout"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Scan {
+    indices: Option<Array>,
+    projection: Projection,
+    filter: Option<RowFilter>,
+    batch_size: usize,
+}
+
+/// Unique identifier for a message within a layout
+pub type MessageId = u16;
+
+#[derive(Debug)]
+pub enum ReadResult {
+    GetMsgs(
+        Vec<(Vec<MessageId>, ByteRange)>,
+        Vec<(Vec<MessageId>, ByteRange)>,
+    ),
+    Batch(Array),
+}
+
+pub enum PlanResult {
+    GetMsg(Vec<(Vec<MessageId>, ByteRange)>),
+    Batch(Vec<MessageId>, Array),
+}
+
+#[derive(Default, Debug)]
+pub struct MessagesCache {
+    cache: HashMap<Vec<MessageId>, Bytes>,
+}
+
+impl MessagesCache {
+    pub fn get(&self, path: &[MessageId]) -> Option<Bytes> {
+        self.cache.get(path).cloned()
+    }
+
+    pub fn set(&mut self, path: Vec<MessageId>, value: Bytes) {
+        self.cache.insert(path, value);
+    }
+}
+
+#[derive(Debug)]
+pub struct RelativeMessageCache {
+    cache_ref: Arc<RwLock<MessagesCache>>,
+    path: Vec<MessageId>,
+    dtype: DType,
+}
+
+impl RelativeMessageCache {
+    pub fn new(dtype: DType, cache_ref: Arc<RwLock<MessagesCache>>, path: Vec<MessageId>) -> Self {
+        Self {
+            cache_ref,
+            path,
+            dtype,
+        }
+    }
+
+    pub fn relative(&self, id: MessageId, dtype: DType) -> Self {
+        let mut new_path = self.path.clone();
+        new_path.push(id);
+        Self::new(dtype, self.cache_ref.clone(), new_path)
+    }
+
+    pub fn get(&self, path: &[MessageId]) -> Option<Bytes> {
+        self.cache_ref.read().unwrap().get(&self.absolute_id(path))
+    }
+
+    pub fn get_dtype(&self) -> DType {
+        self.dtype.clone()
+    }
+
+    pub fn absolute_id(&self, path: &[MessageId]) -> Vec<MessageId> {
+        let mut lookup_key = self.path.clone();
+        lookup_key.extend_from_slice(path);
+        lookup_key
+    }
+}
+
+pub trait Layout: Debug + Send {
+    fn read(&mut self) -> VortexResult<Option<ReadResult>>;
+
+    fn plan(&mut self, scan: Scan) -> VortexResult<Option<PlanResult>>;
```

Comment:
It's meant to be here as an attempt to codify the stats pruning api. However, since it's not implemented I can leave it as a comment and remove dummy impls